### PR TITLE
Update 5.9 with state of swift/main

### LINKED ISSUE #609: Regex: Inconsistent atomic access
In `Regex/Core.swift`, we have the following code:

```swift
    /// The program for execution with the matching engine.
    var loweredProgram: MEProgram {
      if let loweredObject = _loweredProgramStorage as? ProgramBox {
        return loweredObject.value
      }
      let lowered = try! Compiler(tree: tree, compileOptions: compileOptions).emit()
      _stdlib_atomicInitializeARCRef(object: &_loweredProgramStorage, desired: ProgramBox(lowered))
      return lowered
    }
```

This allows regular non-atomic loads of the `_loweredProgramStorage` variable to occur concurrently with atomic updates (cmpxchg) of it. This is explicitly undefined behavior per the Law of Exclusivity:

> Two accesses to the same variable aren't allowed to overlap unless both accesses are reads or both accesses are atomic. -- [SE-0282](https://github.com/apple/swift-evolution/blob/main/proposals/0282-atomics.md)

We _must_ avoid such accesses by converting the nonatomic load into an atomic load instruction:

```swift
    if let ref = _stdlib_atomicLoadARCRef(object: &_loweredProgramStorage) {
      return unsafeDowncast(ref, to: ProgramBox.self)
    }
```

On a secondary note, in case the current thread loses a concurrent update race, this property ought to return the winner program box, not the loser.

### LINKED ISSUE #625: Runtime crash when accessing DSL-composed Regex Match with named capture groups
**Description**
Composing a `Regex` with the DSL using `RegexLiteral`s with named capture groups works, and matching works, but accessing the matches crashes at runtime. 

Using the same patterns in `RegexLiteral` & DSL without the names works fine.

_forum.swift.org: [runtime-crash-regexbuilding-with-regex-literals-containing-named-capture-groups/62539](https://forums.swift.org/t/runtime-crash-regexbuilding-with-regex-literals-containing-named-capture-groups/62539)_

**Steps to reproduce**
Define one or more patterns with named capture groups, and use the DSL to compose a `Regex`.  Match (`whole`, `prefix`, etc.) to a string.  The the pattern matches correctly, but accessing the `Output` causes a runtime crash.

```
enum FrameComponentRegex {
    static let number = /frame \#(?<frame_number>\d+):/
    static let address = /(?<frame_address>\dx\w{16})/
    static let module = /(?<frame_module>\w+\.{0,1}\w+)/
    static let function = /(?<frame_function>.+)/
}

let frame = Regex {
    FrameComponentRegex.number
    One(.whitespace)
    FrameComponentRegex.address
    One(.whitespace)
    FrameComponentRegex.module
    One("`")
    FrameComponentRegex.function
}

let frameString = """
frame apple/swift#10: 0x00007fff2011383a libdispatch.dylib`_dispatch_client_callout + 8
"""

if let match = frameString.wholeMatch(of: frame) {
    print(match) 
/**
Match(anyRegexOutput: _StringProcessing.AnyRegexOutput(
  input: "frame apple/swift#10: 0x00007fff2011383a libdispatch.dylib`_dispatch_client_callout + 8", 
  _elements: [
    _StringProcessing.AnyRegexOutput.ElementRepresentation(optionalDepth: 0, content: Optional((range: Range(Swift.String.Index(_rawBits: 15)..<Swift.String.Index(_rawBits: 4980743)), value: nil)), name: nil, referenceID: nil), 
    _StringProcessing.AnyRegexOutput.ElementRepresentation(optionalDepth: 0, content: Optional((range: Range(Swift.String.Index(_rawBits: 458757)..<Swift.String.Index(_rawBits: 590087)), value: nil)), name: Optional("frame_number"), referenceID: nil), 
    _StringProcessing.AnyRegexOutput.ElementRepresentation(optionalDepth: 0, content: Optional((range: Range(Swift.String.Index(_rawBits: 721159)..<Swift.String.Index(_rawBits: 1900807)), value: nil)), name: Optional("frame_address"), referenceID: nil), 
    _StringProcessing.AnyRegexOutput.ElementRepresentation(optionalDepth: 0, content: Optional((range: Range(Swift.String.Index(_rawBits: 1966343)..<Swift.String.Index(_rawBits: 3080455)), value: nil)), name: Optional("frame_module"), referenceID: nil), 
    _StringProcessing.AnyRegexOutput.ElementRepresentation(optionalDepth: 0, content: Optional((range: Range(Swift.String.Index(_rawBits: 3145733)..<Swift.String.Index(_rawBits: 4980743)), value: nil)), name: Optional("frame_function"), referenceID: nil)
  ]), 
  range: Range(Swift.String.Index(_rawBits: 15)..<Swift.String.Index(_rawBits: 4980743)))
**/
  
let _ = match.output /** Runtime crash
Could not cast value of type 
'(Swift.Substring, Swift.Substring, Swift.Substring, Swift.Substring, Swift.Substring)' (0x7ff848f83d00) 
to 'Swift.Substring' (0x7ff848af00c8).
**/

let (_, number, address, module, function) = match.output /** Compiler exception
Type of expression is ambiguous without more context 
**/
```

**Expected behavior**
It should be possible to access the captures via tuple.  It would be *nice* to also have subscript-based access, but that's a separate issue.

```
enum FrameComponentRegex_Unnamed {
    static let number = /frame \#(\d+):/
    static let address = /(\dx\w{16})/
    static let module = /(\w+\.{0,1}\w+)/
    static let function = /(.+)/
}
if let match = frameString.wholeMatch(of: frame) {
    let (_, number, address, module, function) = match.output
    print("""
    number:     \(number)
    address:    \(address)
    module:     \(module)
    function:   \(function)
    """)
}
/**
number:     10
address:    0x00007fff2011383a
module:     libdispatch.dylib
function:   _dispatch_client_callout + 8
**/
```

**Environment**
- Swift compiler version info `swift-driver version: 1.62.15 Apple Swift version 5.7.1 (swiftlang-5.7.1.135.3 clang-1400.0.29.51)`
- Xcode version info `Version 14.1 (14B47b)`
- Deployment target: `macOS 13.0`

### PR DESCRIPTION
**Pull request by milseman**
* Atomically load the lowered program (#610)

Since we're atomically initializing the compiled program in `Regex.Program`, we need to pair that with an atomic load.

Resolves #609.

* Add tests for line start/end word boundary diffs (#616)

The `default` and `simple` word boundaries have different behaviors at the start and end of strings/lines. These tests validate that we have the correct behavior implemented. Related to issue #613.

* Add tweaks for Android

* Fix documentation typo (#615)

* Fix abstract for Regex.dotMatchesNewlines(_:). (#614)

The old version looks like it was accidentally duplicated from anchorsMatchLineEndings(_:) just below it.

* Remove `RegexConsumer` and fix its dependencies (#617)

* Remove `RegexConsumer` and fix its dependencies

This eliminates the RegexConsumer type and rewrites its users to call through to other, existing functionality on Regex or in the Algorithms implementations. RegexConsumer doesn't take account of the dual subranges required for matching, so it can produce results that are inconsistent with matches(of:) and ranges(of:), which were rewritten earlier.

rdar://102841216

* Remove remaining from-end algorithm methods

This removes methods that are left over from when we were considering from-end algorithms. These aren't tested and may not have the correct semantics, so it's safer to remove them entirely.

* Improve StringProcessing and RegexBuilder documentation (#611)

This includes documentation improvements for core types/methods, RegexBuilder types along with their generated variadic initializers, and adds some curation. It also includes tests of the documentation code samples.

* Set availability for inverted character class test (#621)

This feature depends on running with a Swift 5.7 stdlib, and fails when that isn't available.

* Add type annotations in RegexBuilder tests

These changes work around a change to the way result builders are compiled that removes the ability for result builder closure outputs to affect the overload resolution elsewhere in an expression.

Workarounds for rdar://104881395 and rdar://104645543

* Workaround for fileprivate array issue

A recent compiler change results in fileprivate arrays sometimes not keeping their buffers around long enough. This change avoids that issue by removing the fileprivate annotations from the affected type.

* Fix an issue where named character classes weren't getting converted in the result builder. <rdar://104480703>

* Stop at end of search string in TwoWaySearcher (#631)

When searching for a substring that doesn't exist, it was possible for TwoWaySearcher to advance beyond the end of the search string, causing a crash. This change adds a `limitedBy:` parameter to that index movement, avoiding the invalid movement.

Fixes rdar://105154010

* Correct misspelling in DSL renderer (#627)

vertial -> vertical

rdar://104602317

* Fix output type mismatch with RegexBuilder (#626)

Some regex literals (and presumably other `Regex` instances) lose their output type information when used in a RegexBuilder closure due to the way the concatenating builder calls are overloaded. In particular, any output type with labeled tuples or where the sum of tuple components in the accumulated and new output types is greater than 10 will be ignored.

Regex internals don't make this distinction, however, so there ends up being a mismatch between what a `Regex.Match` instance tries to produce and the output type of the outermost regex. For example, this code results in a crash, because `regex` is a `Regex<Substring>` but the match tries to produce a `(Substring, number: Substring)`:

    let regex = Regex {
        ZeroOrMore(.whitespace)
        /:(?<number>\d+):/
        ZeroOrMore(.whitespace)
    }
    let match = try regex.wholeMatch(in: " :21: ")
    print(match!.output)

To fix this, we add a new `ignoreCapturesInTypedOutput` DSLTree node to mark situations where the output type is discarded. This status is propagated through the capture list into the match's storage, which lets us produce the correct output type. Note that we can't just drop the capture groups when building the compiled program because (1) different parts of the regex might reference the capture group and (2) all capture groups are available if a developer converts the output to `AnyRegexOutput`.

    let anyOutput = AnyRegexOutput(match)
    // anyOutput[1] == "21"
    // anyOutput["number"] == Optional("21")

Fixes #625. rdar://104823356

Note: Linux seems to crash on different tests when the two customTest overloads have `internal` visibility or are called. Switching one of the functions to be generic over a RegexComponent works around the issue.

* Revert "Merge pull request #628 from apple/result_builder_changes_workaround"

This reverts commit 7e059b7bd4e647bc884a71731fcfdc4fffda0700, reversing changes made to 3ca8b134e4c93e8c6d3f7b060b0aefcf06b92039.

* Use `some` syntax in variadics

This supports a type checker fix after the change in how result builder closure parameters are type-checked.

* Type checker workaround: adjust test

* Further refactor to work around type checker regression

* Align availability macro with OS versions (#641)

* Speed up general character class matching (#642)

Short-circuit Character.isASCII checks inside built in character class matching.

Also, make benchmark try a few more times before giving up.

* Test for \s matching CRLF when scalar matching (#648)

* General ascii fast paths for character classes (#644)

General ASCII fast-paths for builtin character classes

* Remove the unsupported `anyScalar` case (#650)

We decided not to support the `anyScalar` character class, which would match a single Unicode scalar regardless of matching mode. However, its representation was still included in the various character class types in the regex engine, leading to unreachable code and unclear requirements when changing or adding new code. This change removes that representation where possible.

The `DSLTree.Atom.CharacterClass` enum is left unchanged, since it is marked `@_spi(RegexBuilder) public`. Any use of that enum case is handled with a `fatalError("Unsupported")`, and it isn't produced on any code path.

---------

### CODE REVIEW

**Comment by milseman** [comment]
@swift-ci please test

File: Sources/VariadicsGenerator/VariadicsGenerator.swift [hunk 1]
```diff
@@ -14,7 +14,7 @@
 import ArgumentParser
 #if os(macOS)
```
Comment (stephentyrone) [inline Sources/VariadicsGenerator/VariadicsGenerator.swift:15]: Seems like this should be `#if os(Darwin)`, even though we're not running it on any non-mac Darwin targets at present.

Comment (stephentyrone) [reply Sources/VariadicsGenerator/VariadicsGenerator.swift:15]: (obviously, if we're gonna fix this, fix it on main first)

File: Sources/_StringProcessing/Unicode/ASCII.swift [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2021-2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+private var _lineFeed: UInt8 { 0x0A }
```
Comment (stephentyrone) [inline Sources/_StringProcessing/Unicode/ASCII.swift:12]: These would maybe be more readable as `UInt8(ascii: "\n")` etc. Mostly a matter of taste, though.

File: Sources/_StringProcessing/Unicode/ASCII.swift [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+extension UInt8 {
+  var _isASCII: Bool { self < 0x80 }
+
+  // TODO: Bitvectors for the below
+
+  /// Assuming we're ASCII, whether we match `\d`
+  var _asciiIsDigit: Bool {
+    assert(_isASCII)
+    return(_0..._9).contains(self)
```
Comment (stephentyrone) [inline Sources/_StringProcessing/Unicode/ASCII.swift:40]: Why do we define `_isASCIINumber` if not to use it here?

File: Sources/_StringProcessing/Unicode/ASCII.swift [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    case _lineFeed, _carriageReturn, _lineTab, _formFeed:
+      return true
+    default:
+      return false
+    }
+  }
+
+  /// Assuming we're ASCII, whether we match `\s`
+  var _asciiIsWhitespace: Bool {
+    assert(_isASCII)
```
Comment (stephentyrone) [inline Sources/_StringProcessing/Unicode/ASCII.swift:62]: I expect that we'd get better codegen out of something like:
```
if self > 0x20 { return false }       // all whitespace is 0x20 or lower
return 0x2e00 &>> self & 1 == 1 // table lookup for whitespace chars
```
Ideally LLVM would be clever enough to do this for us, but it isn't.

File: Sources/_StringProcessing/Unicode/ASCII.swift [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+      return true
+    default:
+      return false
+    }
+  }
+
+  /// Assuming we're ASCII, whether we match `[a-zA-Z]`
+  var _asciiIsLetter: Bool {
+    assert(_isASCII)
+    return (_a..._z).contains(self) || (_A..._Z).contains(self)
```
Comment (stephentyrone) [inline Sources/_StringProcessing/Unicode/ASCII.swift:74]: Check codgen for this; it should be just an AND or OR or BIC, followed by a subtract and compare. If not, we might need to rewrite it that way explicitly.

File: Sources/VariadicsGenerator/VariadicsGenerator.swift (hunk 1 above)
Comment (milseman) [reply Sources/VariadicsGenerator/VariadicsGenerator.swift:15]: @natecook1000  thoughts?

File: Sources/_StringProcessing/Unicode/ASCII.swift (hunk 4 above)
Comment (milseman) [reply Sources/_StringProcessing/Unicode/ASCII.swift:62]: For ASCII character classes, would it be better to have 128-bit bit vectors that we lookup the corresponding ASCII value in?

Comment (stephentyrone) [reply Sources/_StringProcessing/Unicode/ASCII.swift:62]: 128b bit vectors are quite a bit fussier; since there's no 128b bitwise shifts or comparisons on x86 or arm. Where we can get away with 32b or 64b bit vectors (like here), that's usually a win.

File: Sources/_StringProcessing/Unicode/ASCII.swift (hunk 5 above)
Comment (milseman) [reply Sources/_StringProcessing/Unicode/ASCII.swift:74]: Would that be better than a bit vector? 

These improvements are low priority as we also have such overwhelming inefficiencies elsewhere that time spent inside `_quickMatch` is close to negligible. However, I will definitely file something to track future improvements here.

Comment (stephentyrone) [reply Sources/_StringProcessing/Unicode/ASCII.swift:74]: Yes, better than a bit vector.

**Comment by milseman** [comment]
Testing passed (https://github.com/apple/swift/pull/64937). @stephentyrone good to merge to 5.9 and do these fixes on main?

**Review by stephentyrone — APPROVED** [review]
